aws/lambda/log-classifier/src/engine.rs

```diff
@@ -44,3 +46,324 @@ pub fn evaluate_ruleset(ruleset: &RuleSet, log: &Log) -> Option<Match> {
         .flat_map(|rule| evaluate_rule(rule, log))
         .max_by(|a, b| a.rule.priority.cmp(&b.rule.priority))
 }
+
+/// Evaluates a ruleset against all lines of a log, returning matching line numbers.
+///
+/// # Arguments
+///
+/// * `ruleset` - A reference to the RuleSet to evaluate.
+/// * `log` - A reference to the Log to evaluate against.
+///
+/// # Returns
+///
+/// A vector of line numbers that match any rule in the ruleset.
+/// Note: Each rule in ruleset can only match 1 line which is the last line that matches the rule.
+pub fn evaluate_ruleset_all_lines(ruleset: &RuleSet, log: &Log) -> Vec<usize> {
+    let matches: Vec<Match> = ruleset
+        .rules
+        .par_iter()
+        .flat_map(|rule| evaluate_rule(rule, log))
+        .collect();
+    // combine captures into a single vector
+    let mut lines = Vec::new();
+    for m in matches {
+        lines.push(m.line_number);
+    }
+    lines
+}
+
+/// Merges line numbers into chunks based on specified size constraints.
+///
+/// # Arguments
+///
+/// * `log` - A reference to the Log structure containing the lines.
+/// * `line_numbers` - A vector of line numbers to be chunked.
+/// * `min_chunk_size` - The minimum size of each chunk.
+/// * `max_chunk_size` - The maximum size of each chunk.
+///
+/// # Returns
+///
+/// A vector of tuples, where each tuple contains the start and end line numbers of a chunk.
+///
+/// # Panics
+///
+/// This function will panic if:
+/// * `min_chunk_size` is greater than or equal to `max_chunk_size`.
+/// * Any line number is less than 1 or greater than the number of lines in the log.
+fn get_line_number_chunks(
+    log: &Log,
+    line_numbers: Vec<usize>,
+    min_chunk_size: usize,
+    max_chunk_size: usize,
+) -> Vec<(usize, usize)> {
+    // throw user error if min_chunk_size is greater than max_chunk_size
+    if min_chunk_size >= max_chunk_size {
+        panic!("min_chunk_size cannot be greater or equal to max_chunk_size");
+    }
+
+    // check if empty
+    if line_numbers.is_empty() {
+        return Vec::new();
+    }
+
+    // sort line numbers
+    let mut line_numbers = line_numbers;
+    line_numbers.sort();
+
+    // assert length of log is at least last line number and no negative line numbers
+    if *line_numbers.first().unwrap() < 1 || *line_numbers.last().unwrap() > log.lines.len() {
+        panic!("line numbers must be within the range of the log");
+    }
+
+    // merge line numbers into chunks
+    let mut merged_line_numbers = Vec::new();
+    // as we add min_chunk_size / 2 to both sides
+    let effective_max_chunk_size = max_chunk_size - min_chunk_size;
+    let mut chunk_start = 0;
+    let mut chunk_end = 0;
+    for line_number in line_numbers {
+        if chunk_start == 0 {
+            chunk_start = line_number;
+            chunk_end = line_number;
+        } else if line_number - chunk_start <= effective_max_chunk_size {
```

Comment:
I don't think it's correct to compare with `effective_max_chunk_size`, unless you want to create multiple overlapping chunks.

I suggest switching to `line - chunk_end <= max_chunk_size:` and `mid_point` (see my pseudocode):
https://github.com/pytorch/test-infra/pull/5425#discussion_r1676569425

Reply:
yeah you want multiple overlapping chunks. Generally, there isn't a problem with having too many chunks due to costs, but if the chunk is missing context you are sad.